### DISCUSSION TITLE
Avalonia on top of Microsoft.Maui.Graphics

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#7114] [General]
Next year I expect MAUI to pick up more traction for cross-platform UI development (although with Microsoft who really knows). There are also experiments such as https://github.com/dotnet/Microsoft.Maui.Graphics.Controls on the horizon which is rendering controls like Xamarin Forms should have done all along and what Avalonia does today. I expect this is going to create added turbulence in the ecosystem for a while. There will soon be three cross-platform, fully-rendered UI frameworks for .NET: Avalonia, Uno and Maui.Graphics.Controls. While it's great for C# devs to have options -- who would have thought 10 years ago the future would look this good for cross-platform .NET UI's -- I hate to see further fragmentation. It would also be disappointing long-term to see Avalonia loose influence due to Microsoft's direct competition. Avalonia did it first and did it better.

That said I had an idea to better position Avalonia and provide tighter integration with Maui going forward: Swap out the SkiaSharp rendering backend and run Avalonia on top of https://github.com/dotnet/Microsoft.Maui.Graphics. This more tightly integrates Avalonia with Maui which helps from both a marketing and perhaps a technical perspective. It also shouldn't be so difficult considering how clean your rendering interfaces are. 

I've also noticed a clear slowdown with SkiaSharp recently as I expect Microsoft will transition to Microsoft.Maui.Graphics more in the future (although it currently sits on top of SkiaSharp as an abstraction for the most part).

**Comment by danielmayost** [comment]
Avalonia still has a clear advantage that MAUI does not have, Linux!
(MAUI "Community Support" for Linux is not serious)

**Reply by robloo** [reply]
Yes, Avalonia has many, many advantages. I expect it will be a long time before Microsoft offers anything technically that can challenge Avalonia as well (if they ever get there). My concern is Microsoft's marketing resources drowning out Avalonia regardless of technical superiority.

**Reply by mysteryx93** [reply]
I'm kind of fed up with Microsoft. They released WinForms, WPF, UWP, WinUI1, WinUI2, WinUI3, then MAUI -- and after all that, after 20 years, they still haven't been able to deliver a cross-platform UI, and once WinUI3 and MAUI go out of beta, it still won't support Linux. And the truth is, Linux will be growing a lot in the next few years with people migrating away from Windows 11.

From a business perspective, who likes to flip between UI frameworks every few years? By the time I learnt WPF, it was already outdated, but with nothing serious to replace it. I think I'll stick to Avalonia and stop running after the next shiny thing. Better develop applications over an UI framework that will last more than 5 years.

UNO isn't competition for desktop applications, it has UWP limitations which is good for mobile but not suitable for most desktop apps.

I'm not seeing Avalonia getting buried anytime soon. I'm putting my eggs in this basket.

**Reply by robloo** [reply]
@mysteryx93 I think we all share your sentiments. Several in Microsoft wanted to make a true 'Flutter-like' framework about 10 years ago -- before Flutter even existed. They were pushed out of the company and it's why we are where we are today. The good people with the right ideas lost... Microsoft's poor management direction and lack of vision won and it's been all down hill since. A lot of wasted time for everyone for sure.

UWP really was Microsoft's best option to make a cross-platform framework but they failed even on Windows with it. Uno Platform is doing an admirable job making it Universal/cross-platform and I use it myself for mobile where Avalonia currently is lacking. I agree 100% Uno Platform it isn't ready for desktop though (Did a strategic assessment here: https://github.com/robloo/PublicDocs/blob/master/XAMLFrameworkComparison.md).

Avalonia is quite a gem and I hope it goes on to dominate in the future. In terms of architecture it already is the best. Avalonia is what Microsoft should have done right after WPF instead of rewriting to C++, moving to the Windows team (which killed it), and then refusing to take it cross-platform.

I do however foresee concerns with Microsoft finally pushing into this space in the future. It would be a shame for the better product to loose due to marketing resources. Uno Platform, for example, is better known than Avalonia but it is harder to use and has a lot more bugs.

**Reply by jmacato** [reply]
> Several in Microsoft wanted to make a true 'Flutter-like' framework about 10 years ago -- before Flutter even existed.

Interesting! I'd like to know more about this history if there's any data online :) Quite curious haha

**Comment by maxkatz6** [comment]
With higher level of abstraction, the less control we have over rendering. We already have problems with that. 
Not sure if yet another layer will be helpful, at least at current state. 

Also it's definitely possible even now to write your own rendering backend that uses Maui.Graphics.

**Reply by robloo** [reply]
I'm not overly familiar with the issues you might have today -- I know the 3D graphics support has been tricky though. Maui.Graphics is also brand-new so there would be issues. If Maui.Graphics always sits on top of SkiaSharp it would be an unnecessary layer of abstraction and potential performance hit. So I understand your points.

This is more for Avalonia UI to integrate more with MAUI for marketing purposes. A lot of companies will simply go with MAUI because it's backed by Microsoft. If Avalonia could say it's another layer directly on top of MAUI using the graphics subsystem and .NET runtime it just helps sell Avalonia. Honestly, I'm approaching this from a marketing viewpoint for the most part.

**Reply by jmacato** [reply]
yeah i can see that, and i agree somewhat. but we need to know if Maui.Graphics is as capable as Skia for our requirements in the DrawingContext API. I do hope it is though.

**Comment by mysteryx93** [comment]
IMO what is slowing down Avalonia's adoption is not future WinUI3/MAUI, but rather a lack of support and documentation -- although this has improved lately. Some parts are now very well documented, while the rest of the framework remains undocumented. That's what scaring people from looking at how mature the platform actually is.

**Reply by npolyak** [reply]
Concerning the documentation - here is a bunch of code project articles that might help you:

1. [Multiplatform UI Coding with AvaloniaUI in Easy Samples. Part 1 - AvaloniaUI Building Blocks](https://www.codeproject.com/Articles/5308645/Multiplatform-UI-Coding-with-AvaloniaUI-in-Easy-Sa)
2. [Basics of XAML in Easy Samples for Multiplatform Avalonia .NET Framework](https://www.codeproject.com/Articles/5314369/Basics-of-XAML-in-Easy-Samples-for-Multiplatform-A)
3. [Multiplatform Avalonia .NET Framework Programming Basic Concepts in Easy Samples](https://www.codeproject.com/Articles/5311995/Multiplatform-Avalonia-NET-Framework-Programming-B)
4. [Multiplatform Avalonia .NET Framework Programming Advanced Concepts in Easy Samples](https://www.codeproject.com/Articles/5317055/Multiplatform-Avalonia-NET-Framework-Programming-A)

**Reply by mysteryx93** [reply]
There are good articles covering Avalonia concepts; but once we need documentations about more specific and obscure classes, we're on our own.

**Reply by robloo** [reply]
The authoring controls docs need a lot of work. I am still planning to add some controls to Avalonia (SplitButton, etc.) but don't have all the concepts in my head quite yet. I come from a strong WPF/UWP background and the differences are just enough to require a lot of legwork to get started. It would be great to have an example control section and steps to create one. I might add that myself after I go through it the first time.

Separately, @npolyak has a lot of great tutorials. I wonder if some content from these could be pulled into Avalonia docs themselves to help.

**Reply by npolyak** [reply]
Actually that was the plan - to incorporate what I wrote into the tutorials. But that has not been done so far.

**Reply by odalet** [reply]
I also think that the versioning scheme 0.10, 0.11,  can be a bit scary at least to potential corporate customers...

**Reply by odalet** [reply]
Simply replacing this 0. by 1. may incite some "deciding tech leads or architects" to give Avalonia more than a "meh... beta stuff" look.

**Reply by npolyak** [reply]
> Simply replacing this 0. by 1. may incite some "deciding tech leads or architects" to give Avalonia more than a "meh... beta stuff" look.

I completely agree. The product is ready. Microsoft was releasing packages in a much worse state than Avalonia currently. Make it 1.xx.xx

**Reply by jp2masa** [reply]
Last time I saw this being discussed, the idea of the Avalonia team was to start numbering like 11.x for the next major release (IIRC because React did it as well), so I believe it will look more "ready" to those leads and architects really soon.

**Comment by npolyak** [comment]
I have a pretty cynical attitude towards UWP/Xamarin/MAUI. From my point of view they are just poor imitations of WPF/Silverlight products. This is why I prefer Avalonia over Uno.

**Reply by robloo** [reply]
UWP is now much more tolerable than it was when it initially launched -- it is actually built directly from the Silverlight codebase as well. It can also easily be used for desktop-class applications as well as long as you don't run into WinRT system resource bottlenecks with the sandbox. A lot of the design decisions made with UWP that differ from WPF were necessary to get it to work on mobile devices. WPF was only ever designed for desktop. WinRT is where most of the issues were, not with the XAML UI layer. I'll agree with you for Xamarin/MAUI that it was a poor architecture and implementation from the beginning. Now that this is controlled by the .NET Microsoft team they are reorienting the project with MAUI that will make it a direct competitor with Avalonia going forward.

**Reply by npolyak** [reply]
If you check, say Dice.com - there are still only 1-2 UWP jobs for the whole US of A. And this is in spite of the fact that MS had its PR muscle behind it for ~8 years. Perhaps it became more tolerable now, but myself and a lot of other people are still in the mindset that it was done by people who have no clue about WPF concepts and architecture as a bone thrown towards XAML development.

**Reply by robloo** [reply]
I certainly agree. I can also confirm that the people working on it over the past 4 hears or so have no clue about WPF and Xaml. I could show you many github issues where Microsoft wonders what people are taking about and I have to point out that its the same as an old WPF feature, lol. I won't go down the rabbit hole though and I largely agree with you.

**Comment by npolyak** [comment]
One more thing - I would not rely on Microsoft to provide a great multiplatform solution. Perhaps one of the **true** reasons why Silverlight was nixed was that it was undermining MS own platform. This is why I hope that Avalonia can provide the multiplatform solution without any competition from MS for a long time. 

As for why Avalonia's adoption is slow, yes, support and documentation can and should be improved, but the main thing is that there should be more PR and marketing done (preferably by the professionals). Also I believe, some mobile support is a must.

**Reply by mysteryx93** [reply]
JetBrains Rider endorsing Avalonia is a big plus.

Mobile support would really be a HUGE win for Avalonia.

Remains that -- Avalonia is still lacking very basic features. Cannot disable/hide Minimize button? No MessageBox support? What else is missing?

**Reply by npolyak** [reply]
Yes, you can - I have a custom window in NP.Avalonia.Visuals library where you can do whatever you want with the window header. An article is coming explaining the custom window functionality. MessageBox is not needed if you have Window.ShowDialog method.

**Reply by mysteryx93** [reply]
Sure it can be implemented manually in each app -- but often we just need to display simple messages without going through the hassles of reinventing the wheel. Then there's MessageBox.Avalonia.

**Reply by npolyak** [reply]
MessageBox is a bad wheel :) - people tend to use it in a way that would add the Visual functionality into the view models. But it is easy to create if you need it by wrapping ShowDialog(...) method.

**Reply by npolyak** [reply]
I am building NP.Avalonia.Visuals library. Perhaps I'll add some Message Box capabilities to it accessible via non-visual IMessageBox interface. It will take me a bit of time though to add it and to document it.

**Reply by robloo** [reply]
After Sinofsky was removed following Win 8 and the ouster/reassignment of senior management responsible for mobile, Microsoft has embraced cross-platform. This was true first with office team (remember when it only ran on Windows Phone) and then all the web development/PWA tools Microsoft embraced. The mistake was they thought only web apps were the future so let everything else die at that point. That was corrected for .NET starting with the purchase of Xamarin.

If you follow MAUI and .NET development you can see Microsoft is certainly embracing cross-platform solutions. One fundamental issue all along is XAML/UWP responsibility transitioned over to the Windows team away from the .NET developer team (whatever it was called at the time). There was a lot of infighting between these teams, Windows team doesn't have experience with .NET, and the fact the Windows team obviously didn't care about cross-platform seriously crippled XAML. Now that MAUI exists and responsibility is back to the right team -- .NET -- things are changing. Remember it was the .NET team originally responsible for WPF. This team is much more competent, future minded and developer friendly than the Windows team. You will see differences in the way MAUI evolves going forward (assuming David Ortinau's influence fades). 

Bottom line, **don't underestimate what MAUI will become. I see a lot of individuals/companies with the following sentiment: https://github.com/dotnet/Microsoft.Maui.Graphics.Controls/discussions/47#discussioncomment-1792904**

**Reply by npolyak** [reply]
Yes, I think they do want to embrace multiplatform solution, but do it in a very clumsy way. Silverlight was the better way to go than Xamarin (from my point of view). They still do not have a multiplatform version of Visual Studio - Visual Studio for mac is still a completely different product. If they wanted to they could have pushed Silverlight to be the solution. I think they never did primarily because they were scared of offending their own powerful Windows division.

**Reply by npolyak** [reply]
I agree, that MAUI, if properly done can be a great solution - but currently it is not and it remains to be seen if it will really become one. With Microsoft's blind man's buff UI game since about 2012, I am still a bit pessimistic.

**Reply by npolyak** [reply]
And btw, the UWP fiasco, shows that even Microsoft with its powerful PR and marketing team cannot put enough lipstick on a pig for it to become attractive.

**Reply by mysteryx93** [reply]
> lipstick on a pig for it to become attractive

Are you talking about Windows 11?

**Reply by npolyak** [reply]
No, just UWP and the fact that after 8 years of promotion by Microsoft there are only 1-2 UWP jobs on dice for the whole USA. I reserve my judgement about Window 11, but I'll be careful, based on what I heard and switch to it only when I hear some positive reviews from others who are not in Microsoft's pay. So far I only heard from one guy on the codeproject who switched to it and than switched back to Window 10 because of some bugs. That means I'll be waiting at least for the next patch.

**Comment by tyHdev** [comment]
Does anyone know whether Avalonia Android development is/ will be supported on Linux development systems? This is not the case for Xamarin and from what I have been hearing, unlikely to be the case with MAUI.

**Reply by kekekeks** [reply]
.NET 6 should support Android development with Linux now - https://github.com/xamarin/xamarin-android/pull/6613

**Reply by tyHdev** [reply]
Wow, thank you for the quick reply, if true this is excellent news. As a Linux developer, dubious Android support is the only thing making me hesitate about diving head first into Avalonia (which otherwise is by far the most compelling framework for my project requirements).

Is Android support in Avalonia expected to be production-ready in the next 6 months?

**Reply by kekekeks** [reply]
There are plans to improve Android support, but you never know if it works on all devices due to the platform fragmentation.

**Reply by tyHdev** [reply]
For my use case there is a specific set of Android devices I will be prioritizing, but that is definitely something to take into consideration. Thanks for the update!

**Comment by najak3d** [comment]
Our app, iFlyGPS is being ported to Xamarin.Forms, then onto MAUI next year.   We do make fair use of SyncFusion.

I am working hard to push to migrate to Avalonia instead of MAUI.  One stopper that I see is lack of SyncFusion (or similar competitor) support for Avalonia.

I do agree with @robloo  that synergizing with MAUI could be a major PR benefit, which then attracts packages like SyncFusion to offer support on Avalonia, which then gives the green-light to projects like ours, saying "Yes, Avalonia is serious".   Since SyncFusion doesn't consider Avalonia worthy of support - that serves like a red-light to serious projects.

But as all can see the major Controls suppliers are all supporting MAUI.   This makes all the difference to business projects.

Getting in bed with MAUI in whatever way we can, will vastly promote it as a viable alternative to pure-MAUI.   Avalonia could be the "Pixel-Perfect solution" for MAUI.   For starters, could integrate to MAUI very loosely by simply rendering each Avalonia window to a SKia RenderSurface (or Maui.graphics surface).   From MAUI perspective, we'd just be a PictureBox rapidly updating and consuming Mouse/Touch/Keyboard inputs. 

This is the type of integration we have now for Urholonia (Urho.NET + Avalonia).   Each Avalonia window translates to a single UI Sprite in Urho, and it works perfectly nice.   I'd love to see a similar integration with MAUI.  It'll attract a lot of attention.

**Reply by npolyak** [reply]
Yes, MS has bigger PR muscle than an open source group. And yet, I believe, that Avalonia is much better, much more developer friendly and has more potential. As an example of an MS failure check UWP. A lot companies were producing custom controls for UWP, but if you search on dice.com you are likely to see only one UWP job for the whole USA. 
Once Avalonia stabilizes its product for the Web and releases products for Android and IOS mobile devices, it should dominate the market. And I hope that MS, being a great company as they are in every respect but the UI development, will get back to their senses and also put their muscle behind Avalonia. 
Also, from my previous exposure to SynchFusion - it is not the best - I'd recommend Telerik or DevEx instead. What controls do you need from SynchFusion? I am trying to build my own library of various Avalonia controls that I plan to release eventually.

**Reply by najak3d** [reply]
I fully agree with your assessment of Avalonia's superiority.  I'd hate to see this become another Beta vs. VHS scenario, where the inferior technology wins out due to ecosystem/momentum/promotional factors.

For our App, our use of SyncFusion controls are light.  We mostly use SyncFusion for the stylistic gains over raw Xamarin - so we use it for most of the basic control types (panels, trees, list view, buttons, msg box, waitIcon, etc), to harbor these advantages.   I'm being generic here because I'm not the UI guy (thus my inexperience with XAML), so don't have good specifics, other than we are not making heavy use of their stuff.  Most of our controls are custom-built, wit SyncFusion as the foundation for many of our controls.

The absence of Telerik/Devex/SF is a red-light for business apps.  They see this as a barometer for "is it safe to go here".  If the Tools packages don't think it's a good idea - then the business apps simply tend to trust that judgement by default.

Putting Avalonia into MAUI (even if by loose integration) - gets our foot-in-the-door, and gains attention, and more people start asking "why isn't this mainstream MAUI, etc?"   And may even inspire the Controls producers to support Avalonia too.

Just framing Avalonia inside of a MAUI app, would get that attention/promotion.   It's a good idea to purposefully step directly onto their turf, to demand this attention.   When you google MAUI, Avalonia-plugin should come up in the search results.  That's how I see it.

**Reply by npolyak** [reply]
Yes, I agree, that many decision makers prefer to have a 3rd party toolkit for a framework. In my, experience however, all the controls need to be customized whether they are from a 3rd party toolkit or native controls. I prefer to use native controls for WPF unless it is complex enough to build out of native components, e.g. LayoutControl or a Diagram control.

**Comment by wieslawsoltes** [comment]
There are two main issues using Microsoft.Maui.Graphics:

1. https://github.com/dotnet/Microsoft.Maui.Graphics#status
"This is an experimental library; "

2. https://github.com/dotnet/Microsoft.Maui.Graphics#disclaimer
"There is no official support. Use at your own Risk."

If you see the PRs there are many breaking and refactoring changes. The library lacks some basic 2D stuff not to mention 3D integration. We would lose a lot of features using it at current status. There is also no guarantee that it will be ever released.

**Reply by najak3d** [reply]
If a MAUI-based customer renderer were created, I'd imagine we just use SkiaSharp directly -- just to make it easy (which is what Maui.Graphics is anyways).   Then if Maui.Graphics becomes released/mainstream - then porting to it will mostly just be a changing of namespaces.

The integration that Avalonia might consider sooner than later - is the simple one where we just integrate it in the same fashion as we've done with Urho.NET.   Create a custom "AvaloniaControl" in Maui that is essentially just a Skia Rendering surface, which also routes User inputs (mouse, touch, and keyboard) to Avalonia whenever the focus is set onto this control.   That's how we did it for Urholonia; and would imagine the implementation there would be quite similar.

What it buys us is the ability to render Avalonia directly into a Maui app -- either as the full UI (where each Avalonia Window is mapped to one of these surfaces), or it can simply appear anywhere inside the Maui app where they present one of these AvaloniaSurfaces.

It's a fairly easy integration conceptually, but I'm not sure if Maui itself has set up any added roadblocks/obstacles to make it more difficult.

Our app, IFlyGPS, is technically a WPF app, but our whole app is rendered like this - in WPF, the layout is simple -- one PictureBox covers the entire Window, and that's all there is to it...  then we update the bitmap and rapidly blit the result to the PictureBox -- and nobody even knows we're a WPF app.   This same method works seamlessly the same on all 3 platforms that we support-- windows, Android, iOS, with about 97% code re-use.

I would imagine the same scheme would work for Maui with Avalonia.

**Reply by kekekeks** [reply]
Note that Xamarin.Forms integration existed at some point by reusing Avalonia-powered native views.

**Reply by najak3d** [reply]
That is great news.   Avalonia wins again.  Still batting 1000%.

And so porting that to MAUI should be minimal work.  Maui has put a lot of focus on backward compatibility to make existing Xamarin.Forms. solution easy-to-port with minimal work.

**Reply by timahrentlov** [reply]
MAUI is sorely missing a WASM target. Strangely but typical of MS, they seem to be going in the opposite direction instead - using [blazor/razor tech](https://docs.microsoft.com/en-us/mobile-blazor-bindings/) to target the existing MAUI platforms.

So there's a big hole in their MAUI offering which just begs to be filled with something like Avalonia. The perfect Trojan horse.
